src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -207,12 +220,20 @@ entails SolverOptions{..} constraint context hints =
             TypeClassData{ typeClassDependencies } <- case M.lookup className' classesInScope of
               Nothing -> throwError . errorMessage $ UnknownClass className'
               Just tcd -> pure tcd
-            let instances =
-                  [ (substs, tcd)
-                  | tcd <- forClassName (combineContexts context inferred) className' tys''
-                    -- Make sure the type unifies with the type in the type instance definition
-                  , substs <- maybeToList (matches typeClassDependencies tcd tys'')
-                  ]
+            let instances = do
+                  -- TODO: efficiency
+                  chain <- groupBy ((==) `on` tcdChain) $ forClassName (combineContexts context inferred) className' tys''
+                  -- process instances in a chain in index order
+                  let found = for (sortBy (compare `on` tcdIndex) chain) $ \tcd ->
```

Comment:
Hmm, I might need to special-case when `tcdChain = []`.  The processing of any non-chain instances shouldn't interfere with each other.  In this case we wouldn't get the overlapping instances warning.